core/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/BalancedShardsAllocator.java

```diff
@@ -309,6 +321,104 @@ private void balance() {
+                        "rendering the cluster imbalanced again"
+                );
+            }
+
+            sorter.reset(shard.getIndexName());
+            ModelNode[] modelNodes = sorter.modelNodes;
+            final String currentNodeId = shard.currentNodeId();
+            // find currently assigned node
+            ModelNode currentNode = null;
+            for (ModelNode node : modelNodes) {
+                if (node.getNodeId().equals(currentNodeId)) {
+                    currentNode = node;
+                    break;
+                }
+            }
+            assert currentNode != null : "currently assigned node could not be found";
+
+            // balance the shard, if a better node can be found
+            final float currentWeight = sorter.weight(currentNode);
+            final AllocationDeciders deciders = allocation.deciders();
+            final String idxName = shard.getIndexName();
+            Map<String, NodeRebalanceDecision> nodeDecisions = new HashMap<>();
+            Type rebalanceDecisionType = Type.NO;
+            String assignedNodeId = null;
+            for (ModelNode node : modelNodes) {
+                if (node == currentNode) {
+                    continue; // skip over node we're currently allocated to it
+                }
+                final Decision canAllocate = deciders.canAllocate(shard, node.getRoutingNode(), allocation);
+                final float nodeWeight = sorter.weight(node);
+                final boolean betterWeightThanCurrent = nodeWeight <= currentWeight;
+                final boolean rebalanceConditionsMet;
+                final boolean deltaAboveThreshold;
+                final float weightWithShardAdded;
+                if (betterWeightThanCurrent) {
+                    final float currentDelta = absDelta(nodeWeight, currentWeight);
+                    deltaAboveThreshold = lessThan(currentDelta, threshold) == false;
+                    weightWithShardAdded = weight.weightShardAdded(this, node, idxName);
+                    final float proposedDelta = weightWithShardAdded - weight.weightShardRemoved(this, currentNode, idxName);
+                    rebalanceConditionsMet = deltaAboveThreshold && proposedDelta < currentDelta;
+                    if (rebalanceConditionsMet && canAllocate.type().higherThan(rebalanceDecisionType)) {
+                        // rebalance to the node, only will get overwritten if the decision here is to
+                        // THROTTLE and we get a decision with YES on another node
+                        rebalanceDecisionType = canAllocate.type();
+                        assignedNodeId = node.getNodeId();
+                    }
+                } else {
+                    rebalanceConditionsMet = false;
+                    deltaAboveThreshold = false;
+                    weightWithShardAdded = Float.POSITIVE_INFINITY;
```

Comment:
I think it's a little strange to define these three variables as `final` in this for loop, then potentially set them only once, then use them in the map `.put` call, instead I think it'd be better to define them before the `if` check as:

```java
boolean rebalanceConditionsMet = false;
boolean deltaAboveThreshold = false;
float weightWithShardAdded = Float.POSITIVE_INFINITY;
```

and then you can remove this `else` branch entirely.